servicetalk-buffer-api/src/main/java/io/servicetalk/buffer/api/CharSequences.java

```diff
@@ -206,39 +206,81 @@ public static int caseInsensitiveHashCode(CharSequence seq) {
     }
 
     /**
-     * Split a given {@link CharSequence} to separate ones on the given {@code delimiter}.
-     * The returned {@link CharSequence}s are created by invoking the {@link CharSequence#subSequence(int, int)} method
-     * on the main one.
+     * Split a given {@link #newAsciiString(Buffer) AsciiString} to separate ones on the given {@code delimiter}.
      *
+     * Trimming white-space before and after each token can be controlled by the {@code trim} flag
      * This method has no support for regex.
      *
-     * @param input The initial {@link CharSequence} to split, this experiences no side effects
-     * @param delimiter The delimiter character
+     * @param input The initial {@link CharSequence} to split, this experiences no side effects.
+     * @param delimiter The delimiter character.
+     * @param trim Flag to control whether the individual items must be trimmed.
      * @return a {@link List} of {@link CharSequence} subsequences of the input with the separated values
      */
-    public static List<CharSequence> split(final CharSequence input, final char delimiter) {
+    public static List<CharSequence> split(final CharSequence input, final char delimiter, final boolean trim) {
         if (input.length() == 0) {
             return emptyList();
         }
 
+        boolean isAscii = isAsciiString(input);
+        List<CharSequence> result = new ArrayList<>(4);
+
         int startIndex = 0;
-        List<CharSequence> result = new ArrayList<>();
-        for (int i = 0; i < input.length(); i++) {
-            if (input.charAt(i) == delimiter) {
-                if ((i - startIndex) > 0) {
-                    result.add(input.subSequence(startIndex, i));
+        int endIndex = -1;
+
+        if (trim) {
+            startIndex = -1;
+            for (int i = 0; i < input.length(); i++) {
+                char c = input.charAt(i);
+                if (c != ' ' && c != delimiter) {
+                    endIndex = i + 1;
                 }
 
-                startIndex = i + 1;
+                if (startIndex == -1 && c != ' ' && c != delimiter) {
+                    startIndex = i;
+                } else if (c == delimiter) {
+                    if (endIndex > startIndex) {
+                        result.add(subsequence(isAscii, input, startIndex, endIndex));
+                    } else {
+                        result.add(isAscii ? newAsciiString("") : "");
+                    }
+
+                    startIndex = -1;
+                    endIndex = -1;
+                }
+            }
+        } else {
+            for (int i = 0; i < input.length(); i++) {
+                char c = input.charAt(i);
+                endIndex = i;
```

Comment:
I think we have an issue if there are consecutive delimiters. Here is a test case to add:

```java
assertThat(split(",,,", ',', false), empty());
```

IIUC we should only need a the following state:
- `i` -> current index
- `startIndex` -> last non-delimiter `char` (reset after each subsequence)

Reply:
I think the same issue applies for the `trim` path too:
```java
assertThat(split(",,,", ',', true), empty());
```

Reply:
This is confusing. Even `String#split` has different outcomes which I find surprising.

- ` " ,,,, ".split(",") => [ , , , ,  ]`
- ` ",,,, ".split(",") => [, , , ,  ]`
- `" ,,,,".split(",") => [ ]`
- `" ,,, ,".split(",") => [ , , ,  ]`
- `" , , , , ".split(",") => [ ,  ,  ,  ,  ]`

Python on the other hand, is more predictable ie. WYSIWYG
eg.

- `" ,,, ,".split(",") => [' ', '', '', ' ', '']`

My expectation is the same; that if you have N delimiters in a string, you want N+1 elements in the result set, regardless if they are empty. Otherwise its surprising to the user for this `split(",,,", ',', _)` to be equal to `split("", ',', _)`.

With the trim=true & del=' ', that is a special case, where I chose to give priority to the del, thus these two should behave similarly

- ` ",,,,".split(",") => [ , , , , ]`
- ` "    ".split("") => [ , , , , ]`

Reply:
🤯 interesting discussion.
Ideally, we should be consistent with JDK behavior of `String#split`, but I'm agreed that it behaves with lots of confusion. 
Let's merge the current version, because it fixes the bug. And we can follow-up here later. Anyway, usage of the `split` method is limited atm to compression headers only.